dm/master/shardddl/pessimist.go

```diff
@@ -0,0 +1,375 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shardddl
+
+import (
+	"context"
+	"sync"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/shardddl/pessimism"
+)
+
+// Pessimist used to coordinate the shard DDL migration in pessimism mode.
+type Pessimist struct {
+	mu sync.Mutex
+
+	logger log.Logger
+
+	closed bool
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	cli *clientv3.Client
+	lk  *pessimism.LockKeeper
+
+	// sources used to get all sources relative to the give task.
+	sources func(task string) []string
+}
+
+// NewPessimist creates a new Pessimist instance.
+func NewPessimist(pLogger *log.Logger, sources func(task string) []string) *Pessimist {
+	return &Pessimist{
+		logger:  pLogger.WithFields(zap.String("component", "shard DDL pessimist")),
+		closed:  true, // mark as closed before started.
+		lk:      pessimism.NewLockKeeper(),
+		sources: sources,
+	}
+}
+
+// Start starts the shard DDL coordination in pessimism mode.
+func (p *Pessimist) Start(pCtx context.Context, etcdCli *clientv3.Client) error {
+	p.logger.Info("the shard DDL pessimist is starting")
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.lk.Clear() // clear all previous locks to support re-Start.
+
+	// get the history shard DDL info.
+	ifm, rev, err := pessimism.GetAllInfo(etcdCli)
+	if err != nil {
+		return err
+	}
+	p.logger.Info("get history shard DDL info", zap.Reflect("info", ifm), zap.Int64("revision", rev))
+
+	// get the history shard DDL lock operation.
+	// the newly operations after this GET will be received through the WATCH with `rev`,
+	// and call `Lock.MarkDone` multiple times is fine.
+	opm, rev2, err := pessimism.GetAllOperations(etcdCli)
+	if err != nil {
+		return err
+	}
+	p.logger.Info("get history shard DDL lock operation", zap.Reflect("operation", opm), zap.Int64("revision", rev2))
+
+	// recover the shard DDL lock based on history shard DDL info & lock operation.
+	err = p.recoverLocks(ifm, opm)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(pCtx)
+
+	// watch for the shard DDL info and handle them.
+	infoCh := make(chan pessimism.Info, 10)
+	p.wg.Add(2)
+	go func() {
+		defer func() {
+			p.wg.Done()
+			close(infoCh)
+		}()
+		pessimism.WatchInfoPut(ctx, etcdCli, rev, infoCh)
+	}()
+	go func() {
+		defer p.wg.Done()
+		p.handleInfoPut(ctx, infoCh)
+	}()
+
+	// watch for the shard DDL lock operation and handle them.
+	opCh := make(chan pessimism.Operation, 10)
+	p.wg.Add(2)
+	go func() {
+		defer func() {
+			p.wg.Done()
+			close(opCh)
+		}()
+		pessimism.WatchOperationPut(ctx, etcdCli, "", "", rev, opCh)
```

Comment:
`rev` can work correctly because `rev < rev2`, but `rev2 is more readable, updated in https://github.com/pingcap/dm/pull/456/commits/3ae58b6dfc51b259a7450742091c1408b919b659.